pkg/asset/installconfig/aws/metadata.go

```diff
@@ -74,6 +75,21 @@ func (m *Metadata) AvailabilityZones(ctx context.Context) ([]string, error) {
 	return m.availabilityZones, nil
 }
 
+// EdgeSubnets retrieves subnet metadata indexed by subnet ID, for
+// subnets that the cloud-provider logic considers to be edge
+// (i.e. Local Zone).
+func (m *Metadata) EdgeSubnets(ctx context.Context) (map[string]Subnet, error) {
+	m.mutex.Lock()
```

Comment:
Regarding the mutex locking, I see this method is doing what all other methods are doing, but wouldn't it be better to grab the mutex lock inside populateSubnets() and remove it from all the functions that are calling it.

Reply:
That's a good point. IMHO sharing this mutex with the session is not a better approach in this case. The `?Subnets()` functions are called from different points with the goal of getting the list of subnets that should be populated at once by `populateSubnets()`. The mutex `m.mutex` is being used also by `Session()` and called from many points from the code. Both should have different goals, although `?Subnets()` (`populateSubnets()`) also consumes the session.

I am introducing a mutex for `populateSubnets()`, used by  `?Subnets()`s, and setting a flag `subnetsPopulated` to avoid calling the `subnets()` many times. It also will help to prevent, eventually, getting into the case of returning empty `EdgeSubnets` when the `(Public||Private)Subnets` has been populated. I think the intention here is to check if the discovery was already been called.

Let me know your thoughts, @sadasu and @patrickdillon .

Reference [patch](patch).

[patch]:https://github.com/openshift/installer/compare/ee094569c1e9b59a55d694f6d83811e57c1c8b87..95bc39b33543d489b20c2da0cf3e38ef6f10dc3f#diff-60605f1bbbaa381ef60e730158a166623607b32eae21f481509e098bd24ff5ba